topics/integrate-in-existing-app.md

```diff
@@ -1,242 +1,423 @@
-                commandView { goToNextScreen() }
-            } catch (ex: LoginException) {
-                commandView {
-                    displayLoading(false)
-                    displayError(ex.code)
-                }
-            }
-        }
-    }
-}
-```
+    ![Greeting from the shared module](shared-module-greeting.png)
 
-Note the `commandView` and `lastCommand` mechanism, which allows a view to detach and re-attach, for example for configuration changes on Android.
+### Make the business logic cross platform
 
-#### MVI for declarative UI frameworks {initial-collapse-state="collapsed"}
+You can now extract the business logic code to the KMM shared module and make it platform independent. This is necessary for reusing it for both Android and iOS.
+ 
+1. Move the business logic code `com.jetbrains.simplelogin.androidapp.data` from the `app` directory to the `com.jetbrains.simplelogin.shared` package in the `shared/src/commonMain` directory.
+   You can drag and drop the package or refactor it by moving everything from one directory to another.
+   
+    ![Drag and drop the package with the business logic code](moving-business-logic.png){width=350}
 
-_Model-View-Intent_ (_MVI_) is the natural evolution of MVP when working with declarative UI frameworks (although it also works with legacy UI frameworks). 
-It is therefore recommended to be used with Swift UI or Jetpack Compose.
+2. When Android Studio asks what you'd like to do, select to move the package, and then approve refactoring.
 
-In MVI, the entire UI structure is described in one tree. Here is an example of a simple MVI presenter:
+    ![Refactor the buiness logic package](refactor-business-logic-package.png){width=500}
 
-```kotlin
-class LoginPresenter {
-    sealed class Model {
-        object Form : Model()
-        object Loading : Model()
-        data class Error(val code: Int): Model()
-        object GoToNextScreen : Model()
-    }
-    interface View {
-        fun displayModel(model: Model)
-    }
+3. Ignore all warnings about platform-dependent code and click **Continue**.
 
-    private var view: View? = null
-    private var lastModel: Model = Model.Form
-    private fun displayModel(model: Model) {
-        lastModel = model
-        view?.displayModel(model)
-    }
+    ![Warnings about platform-dependent code](warnings-android-specific-code.png){width=450}
 
-    fun attach(view: View) {
-        this.view = view.apply { displayModel(lastModel) }
-    }
-    fun detach() { this.view = null }
+4. Remove Android-specific code by replacing it with cross-platform Kotlin code or connecting to Android-specific APIs using [`expect` and `actual` declarations](connect-to-platform-specific-apis.md).
 
-    sealed class Intent {
-        data class Login(val username: String, val password: String)
-            : Intent()
-    }
+#### Replace Android-specific code with cross-platform code {initial-collapse-state="collapsed"}
```

Comment:
If these steps are required, maybe you should:
- Put a colon to the end of the previous sentence
- Explain the steps explicitly, like "do the following" or smth like that